diff -aru nx-X11.orig/programs/Xserver/hw/nxagent/Window.c nx-X11/programs/Xserver/hw/nxagent/Window.c
--- nx-X11.orig/programs/Xserver/hw/nxagent/Window.c 2010-02-02 13:31:50.000000000 +0100
+++ nx-X11/programs/Xserver/hw/nxagent/Window.c 2010-09-20 16:00:09.000000000 +0200
@@ -372,6 +372,13 @@
                                                   visual,
                                                   mask, &attributes);
 
+  
+  Atom clientWid = MakeAtom("_QVD_CLIENT_WID", strlen("_QVD_CLIENT_WID"), True);
+  ChangeWindowProperty(pWin, clientWid, XA_WINDOW, 32, PropModeReplace,
+	/* len */ 1, 
+	/* value */ (unsigned char*) &(nxagentWindow(pWin)),
+	/* sendevent */ 1);
+
   nxagentWindowPriv(pWin) -> isMapped = 0;
   nxagentWindowPriv(pWin) -> isRedirected = 0;
 
@@ -2428,7 +2435,9 @@
        * Map and raise the input window.
        */
 
+/* oh no you don't :)
       XMapWindow(nxagentDisplay, nxagentInputWindows[pScreen->myNum]);
+*/
 
       /*
        * At reconnection the Input Window is
