---
 Makefile |   36 ++++++++++++++++++++++++++++++++++++
 1 file changed, 36 insertions(+)

Index: nspluginwrapper-1.2.2/Makefile
===================================================================
--- nspluginwrapper-1.2.2.orig/Makefile	2010-08-23 08:00:54.000000000 +0200
+++ nspluginwrapper-1.2.2/Makefile	2010-08-23 08:01:20.000000000 +0200
@@ -79,11 +79,13 @@
 ARCH_32 = $(ARCH)
 ifeq ($(biarch), yes)
 ARCH_32 = $(TARGET_ARCH)
+ifeq (,$(USE_SYSTEM_LSB))
 LSB_LIBS = $(LSB_OBJ_DIR)/libc.so $(LSB_OBJ_DIR)/libgcc_s_32.so
 LSB_LIBS += $(LSB_CORE_STUBS:%=$(LSB_OBJ_DIR)/%.so)
 LSB_LIBS += $(LSB_CORE_STATIC_STUBS:%=$(LSB_OBJ_DIR)/%.a)
 LSB_LIBS += $(LSB_DESKTOP_STUBS:%=$(LSB_OBJ_DIR)/%.so)
 endif
+endif
 
 LSB_TOP_DIR = $(SRC_PATH)/lsb-build
 LSB_INC_DIR = $(LSB_TOP_DIR)/headers
@@ -119,7 +121,11 @@
 npviewer_CFLAGS += -I$(LSB_INC_DIR)
 npviewer_CFLAGS += -I$(LSB_INC_DIR)/glib-2.0
 npviewer_CFLAGS += -I$(LSB_INC_DIR)/gtk-2.0
+ifeq (,$(USE_SYSTEM_LSB))
 npviewer_LDFLAGS = $(LDFLAGS_32) -L$(LSB_OBJ_DIR)
+else
+npviewer_LDFLAGS = $(LDFLAGS_32) -L/usr/lib32/
+endif
 npviewer_LDFLAGS += -lgtk-x11-2.0 -lgdk-x11-2.0 -lgobject-2.0 -ldl -lglib-2.0 -lgthread-2.0 -lX11 -lXt
 else
 npviewer_CFLAGS += $(GTK_CFLAGS)
@@ -157,7 +163,11 @@
 libxpcom_CFLAGS  = $(PIC_CFLAGS)
 ifeq ($(biarch),yes)
 libxpcom_CFLAGS += -I$(LSB_INC_DIR)
+ifeq (,$(USE_SYSTEM_LSB))
 libxpcom_LDFLAGS = $(LDFLAGS_32) -L$(LSB_OBJ_DIR)
+else
+libxpcom_LDFLAGS = $(LDFLAGS_32) -L/usr/lib32/
+endif
 endif
 
 libnoxshm_LIBRARY = libnoxshm.so
@@ -167,7 +177,11 @@
 libnoxshm_CFLAGS  = $(PIC_CFLAGS)
 ifeq ($(biarch),yes)
 libnoxshm_CFLAGS += -I$(LSB_INC_DIR)
+ifeq (,$(USE_SYSTEM_LSB))
 libnoxshm_LDFLAGS = $(LDFLAGS_32) -L$(LSB_OBJ_DIR)
+else
+libxpcom_LDFLAGS = $(LDFLAGS_32) -L/usr/lib32/
+endif
 endif
 
 npconfig_PROGRAM = npconfig
@@ -235,6 +249,7 @@
 FILES		+= $(addprefix $(LSB_TOP_DIR)/headers/,$(shell cat $(LSB_TOP_DIR)/headers/core_filelist))
 FILES		+= $(LSB_TOP_DIR)/headers/desktop_filelist
 FILES		+= $(addprefix $(LSB_TOP_DIR)/headers/,$(shell cat $(LSB_TOP_DIR)/headers/desktop_filelist))
+ifeq (,$(USE_SYSTEM_LSB))
 FILES		+= $(LSB_SRC_DIR)/LibNameMap.txt
 FILES		+= $(LSB_SRC_DIR)/core_filelist
 FILES		+= $(LSB_SRC_DIR)/core_static_filelist
@@ -244,6 +259,7 @@
 FILES		+= $(patsubst %,$(LSB_SRC_DIR)/%.c,$(LSB_CORE_STATIC_STUBS))
 FILES		+= $(patsubst %,$(LSB_SRC_DIR)/%.c,$(LSB_DESKTOP_STUBS))
 FILES		+= $(patsubst %,$(LSB_SRC_DIR)/%.Version,$(LSB_DESKTOP_STUBS))
+endif
 
 all: $(TARGETS)
 
@@ -378,8 +394,13 @@
 npwrapper-%.os: $(SRC_PATH)/src/%.c
 	$(CC) -o $@ -c $< $(PIC_CFLAGS) $(CPPFLAGS) $(npwrapper_CFLAGS) -DBUILD_WRAPPER
 
+ifeq (,$(USE_SYSTEM_LSB))
 $(npviewer_PROGRAM): $(npviewer_OBJECTS) $(npviewer_MAPFILE) $(LSB_OBJ_DIR) $(LSB_LIBS)
 	$(CC) $(LDFLAGS_32) -o $@ $(npviewer_OBJECTS) $(npviewer_LDFLAGS)
+else
+$(npviewer_PROGRAM): $(npviewer_OBJECTS) $(npviewer_MAPFILE)
+	$(CC) $(LDFLAGS_32) -o $@ $(npviewer_OBJECTS) $(npviewer_LDFLAGS)
+endif
 
 npviewer-%.o: $(SRC_PATH)/src/%.c
 	$(CC) $(CFLAGS_32) -o $@ -c $< $(CPPFLAGS) $(npviewer_CFLAGS) -DBUILD_VIEWER
@@ -387,22 +408,37 @@
 npviewer-%.o: $(SRC_PATH)/src/%.cpp
 	$(CXX) $(CFLAGS_32) -o $@ -c $< $(CPPFLAGS) $(npviewer_CFLAGS) -DBUILD_VIEWER
 
+ifeq (,$(USE_SYSTEM_LSB))
 $(npplayer_PROGRAM): $(npplayer_OBJECTS) $(npplayer_MAPFILE) $(LSB_OBJ_DIR) $(LSB_LIBS)
 	$(CC) $(LDFLAGS) -o $@ $(npplayer_OBJECTS) $(npplayer_LDFLAGS)
+else
+$(npplayer_PROGRAM): $(npplayer_OBJECTS) $(npplayer_MAPFILE)
+	$(CC) $(LDFLAGS) -o $@ $(npplayer_OBJECTS) $(npplayer_LDFLAGS)
+endif
 
 npplayer-%.o: $(SRC_PATH)/src/%.c
 	$(CC) $(CFLAGS) -o $@ -c $< $(CPPFLAGS) $(npplayer_CFLAGS) -DBUILD_PLAYER
 npplayer-%.o: $(SRC_PATH)/src/tidy/%.c
 	$(CC) $(CFLAGS) -o $@ -c $< $(CPPFLAGS) $(npplayer_CFLAGS) -DBUILD_PLAYER
 
+ifeq (,$(USE_SYSTEM_LSB))
 $(libxpcom_LIBRARY): $(libxpcom_OBJECTS) $(LSB_OBJ_DIR) $(LSB_LIBS)
 	$(CC) $(LDFLAGS_32) $(DSO_LDFLAGS) -o $@ $(libxpcom_OBJECTS) $(libxpcom_LDFLAGS) -Wl,$(LD_soname),libxpcom.so
+else
+$(libxpcom_LIBRARY): $(libxpcom_OBJECTS)
+	$(CC) $(LDFLAGS_32) $(DSO_LDFLAGS) -o $@ $(libxpcom_OBJECTS) $(libxpcom_LDFLAGS) -Wl,$(LD_soname),libxpcom.so
+endif
 
 libxpcom-%.o: $(SRC_PATH)/src/%.c
 	$(CC) $(CFLAGS_32) -o $@ -c $< $(CPPFLAGS) $(libxpcom_CFLAGS) -DBUILD_XPCOM
 
+ifeq (,$(USE_SYSTEM_LSB))
 $(libnoxshm_LIBRARY): $(libnoxshm_OBJECTS) $(LSB_OBJ_DIR) $(LSB_LIBS)
 	$(CC) $(LDFLAGS_32) $(DSO_LDFLAGS) -o $@ $(libnoxshm_OBJECTS) $(libnoxshm_LDFLAGS) -Wl,$(LD_soname),libnoxshm.so
+else
+$(libnoxshm_LIBRARY): $(libnoxshm_OBJECTS)
+	$(CC) $(LDFLAGS_32) $(DSO_LDFLAGS) -o $@ $(libnoxshm_OBJECTS) $(libnoxshm_LDFLAGS) -Wl,$(LD_soname),libnoxshm.so
+endif
 
 libnoxshm-%.o: $(SRC_PATH)/src/%.c
 	$(CC) $(CFLAGS_32) -o $@ -c $< $(CPPFLAGS) $(libnoxshm_CFLAGS)
