diff -Naur nx-orig/nxcomp/Loop.cpp nx-new/nxcomp/Loop.cpp
--- nx-orig/nxcomp/Loop.cpp	2008-04-14 15:39:06.000000000 +0200
+++ nx-new/nxcomp/Loop.cpp	2009-11-25 18:15:12.000000000 +0100
@@ -10979,7 +10979,7 @@
                                       packMethod == PACK_LOSSLESS ||
                                           packMethod == PACK_ADAPTIVE)
   {
-    char *dash = rindex(opt, '-');
+    const char *dash = rindex(opt, '-');
 
     if (dash != NULL && strlen(dash) == 2 &&
             *(dash + 1) >= '0' && *(dash + 1) <= '9')
