--- Frame.pm	2014-02-06 21:38:11.012845075 +0100
+++ Frame.pm.new	2014-02-06 21:37:13.472561609 +0100
@@ -95,28 +95,28 @@
         
 
         ###############################
-        $settings_sizer->Add( Wx::StaticText->new($settings_panel, -1, "Connection"), 0, wxALL, 5);
+        $settings_sizer->Add( Wx::StaticText->new($settings_panel, -1, "Conexion"), 0, wxALL, 5);
         $settings_sizer->Add( Wx::StaticLine->new($settings_panel, -1, wxDefaultPosition, wxDefaultSize, wxLI_HORIZONTAL, "line"), 0, wxEXPAND | wxLEFT | wxRIGHT, 5 );
 
 
-        $self->{audio} = Wx::CheckBox->new($settings_panel, -1, "Enable audio", wxDefaultPosition, wxDefaultSize, 0, wxDefaultValidator, "checkBox");
+        $self->{audio} = Wx::CheckBox->new($settings_panel, -1, "Habilitar sonido", wxDefaultPosition, wxDefaultSize, 0, wxDefaultValidator, "checkBox");
         $self->{audio}->SetValue( core_cfg("client.audio.enable" ) );
         $settings_sizer->Add($self->{audio});
 
-        $self->{printing} = Wx::CheckBox->new($settings_panel, -1, "Enable printing", wxDefaultPosition, wxDefaultSize, 0, wxDefaultValidator, "checkBox");
+        $self->{printing} = Wx::CheckBox->new($settings_panel, -1, "Habilitar Impresion", wxDefaultPosition, wxDefaultSize, 0, wxDefaultValidator, "checkBox");
         $self->{printing}->SetValue( core_cfg("client.printing.enable" ) );
         $settings_sizer->Add($self->{printing});
 
-        $self->{forwarding} = Wx::CheckBox->new($settings_panel, -1, "Enable port forwarding", wxDefaultPosition, wxDefaultSize, 0, wxDefaultValidator, "checkBox");
+        $self->{forwarding} = Wx::CheckBox->new($settings_panel, -1, "Habilitar reenvio de puertos", wxDefaultPosition, wxDefaultSize, 0, wxDefaultValidator, "checkBox");
         $self->{forwarding}->SetValue( core_cfg("client.slave.enable" ) );
         $settings_sizer->Add($self->{forwarding});        
         $settings_sizer->AddSpacer(5);
 
         ###############################
-        $settings_sizer->Add( Wx::StaticText->new($settings_panel, -1, "Screen"), 0, wxALL, 5);
+        $settings_sizer->Add( Wx::StaticText->new($settings_panel, -1, "Pantalla"), 0, wxALL, 5);
         $settings_sizer->Add( Wx::StaticLine->new($settings_panel, -1, wxDefaultPosition, wxDefaultSize, wxLI_HORIZONTAL, "line"), 0, wxEXPAND | wxLEFT | wxRIGHT, 5 );
 
-        $self->{fullscreen} = Wx::CheckBox->new($settings_panel, -1, "Full screen");
+        $self->{fullscreen} = Wx::CheckBox->new($settings_panel, -1, "Pantalla completa");
         $self->{fullscreen}->SetValue( core_cfg("client.fullscreen" ) );
         $settings_sizer->Add($self->{fullscreen}, 0, wxALL, 5);
         
@@ -132,7 +132,7 @@
     my $grid_sizer = Wx::GridSizer->new(1, 2, 0, 0);
     $ver_sizer->Add($grid_sizer, 1, wxALL|wxEXPAND, 20);
 
-    $grid_sizer->Add(Wx::StaticText->new($panel, -1, "User"), 0, wxALL, 5);
+    $grid_sizer->Add(Wx::StaticText->new($panel, -1, "Usuario"), 0, wxALL, 5);
     $self->{username} = Wx::TextCtrl->new($panel, -1, core_cfg('client.remember_username') ?  core_cfg('client.user.name') : "");
     $grid_sizer->Add($self->{username}, 1, wxALL|wxEXPAND, 5);
 
@@ -141,20 +141,20 @@
     $grid_sizer->Add($self->{password}, 0, wxALL|wxEXPAND, 5);
 
     if (core_cfg('client.show.remember_password')) {
-        $grid_sizer->Add(Wx::StaticText->new($panel, -1, "Remember password"), 0, wxALL, 5);
+        $grid_sizer->Add(Wx::StaticText->new($panel, -1, "Recordar password"), 0, wxALL, 5);
         $self->{remember_pass} = Wx::CheckBox->new ($panel, -1, '', wxDefaultPosition);
         $self->{remember_pass}->SetValue(core_cfg('client.remember_password') ? 1 : 0);
         $grid_sizer->Add($self->{remember_pass}, 1, wxALL, 5);
     }
 
     if (!core_cfg('client.force.host.name', 0)) {
-        $grid_sizer->Add(Wx::StaticText->new($panel, -1, "Server"), 0, wxALL, 5);
+        $grid_sizer->Add(Wx::StaticText->new($panel, -1, "Servidor"), 0, wxALL, 5);
         $self->{host} = Wx::TextCtrl->new($panel, -1, core_cfg('client.host.name'));
         $grid_sizer->Add($self->{host}, 1, wxALL|wxEXPAND, 5);
     }
 
     if (!core_cfg('client.force.link', 0)) {
-        $grid_sizer->Add(Wx::StaticText->new($panel, -1, "Connection type"), 0, wxALL, 5);             
+        $grid_sizer->Add(Wx::StaticText->new($panel, -1, "Tipo de enlace"), 0, wxALL, 5);             
         my @link_options = ("Local", "ADSL", "Modem");
         $self->{link} = Wx::Choice->new($panel, -1);
         $grid_sizer->Add($self->{link}, 1, wxALL|wxEXPAND, 5);
@@ -176,7 +176,7 @@
         $self->{link}->Select($link_select);
     }
 
-    $grid_sizer->Add(Wx::StaticText->new($panel, -1, "Kill current VM"), 0, wxALL, 5);
+    $grid_sizer->Add(Wx::StaticText->new($panel, -1, "Reinicio de sesion"), 0, wxALL, 5);
     $self->{kill_vm} = Wx::CheckBox->new ($panel, -1, '', wxDefaultPosition);
     $grid_sizer->Add($self->{kill_vm});
 
@@ -209,7 +209,7 @@
     }
 
     # port goes here!
-    $self->{connect_button} = Wx::Button->new($panel, -1, "Connect");
+    $self->{connect_button} = Wx::Button->new($panel, -1, "Conectar");
     $ver_sizer->Add($self->{connect_button}, 0, wxLEFT|wxRIGHT|wxBOTTOM|wxEXPAND, 20);
     $self->{connect_button}->SetDefault;
 
