diff -Nurp vnc4-4.1.1+xorg4.3.0/unix/xc/programs/Xserver/hw/xfree86/common/compiler.h vnc4-4.1.1+xorg4.3.0-armhf/unix/xc/programs/Xserver/hw/xfree86/common/compiler.h
--- vnc4-4.1.1+xorg4.3.0/unix/xc/programs/Xserver/hw/xfree86/common/compiler.h	2012-09-28 16:48:44.312671616 -0700
+++ vnc4-4.1.1+xorg4.3.0-armhf/unix/xc/programs/Xserver/hw/xfree86/common/compiler.h	2013-01-10 15:42:16.845010046 -0800
@@ -86,8 +86,8 @@ extern int ffs(unsigned long);
 # endif
 
 # if defined(NO_INLINE) || defined(DO_PROTOTYPES)
-#  if !defined(__arm__)
-#   if !defined(__sparc__) && !defined(__arm32__) \
+
+#  if !defined(__sparc__) && !defined(__arm32__) \
       && !(defined(__alpha__) && defined(linux))
 
 extern void outb(unsigned short, unsigned char);
@@ -97,7 +97,7 @@ extern unsigned int inb(unsigned short);
 extern unsigned int inw(unsigned short);
 extern unsigned int inl(unsigned short);
 
-#   else /* __sparc__,  __arm32__, __alpha__*/
+#  else /* __sparc__,  __arm32__, __alpha__*/
 
 extern void outb(unsigned long, unsigned char);
 extern void outw(unsigned long, unsigned short);
@@ -106,8 +106,8 @@ extern unsigned int inb(unsigned long);
 extern unsigned int inw(unsigned long);
 extern unsigned int inl(unsigned long);
 
-#   endif /* __sparc__,  __arm32__, __alpha__ */
-#  endif /* __arm__ */
+#  endif /* __sparc__,  __arm32__, __alpha__ */
+
 extern unsigned long ldq_u(unsigned long *);
 extern unsigned long ldl_u(unsigned int *);
 extern unsigned long ldw_u(unsigned short *);
@@ -857,6 +857,7 @@ static __inline__ void stw_u(unsigned lo
 
 unsigned int IOPortBase;  /* Memory mapped I/O port area */
 
+#    if defined(__mips__)
 static __inline__ void
 outb(unsigned PORT_SIZE port, unsigned char val)
 {
@@ -893,8 +894,6 @@ inl(unsigned PORT_SIZE port)
 	return *(volatile unsigned int*)(((unsigned PORT_SIZE)(port))+IOPortBase);
 }
 
-
-#    if defined(__mips__)
 static __inline__ unsigned long ldq_u(unsigned long * r11)
 {
 	unsigned long r1;
@@ -1244,6 +1243,35 @@ inl(unsigned short port)
 #    define mem_barrier()	eieio()
 #    define write_mem_barrier()	eieio()
 
+#   elif defined(__arm__) && defined(__linux__)
+
+/* for Linux on ARM, we use the LIBC inx/outx routines */
+/* note that the appropriate setup via "ioperm" needs to be done */
+/*  *before* any inx/outx is done. */
+#include <sys/io.h>
+
+static __inline__ void
+xf_outb(unsigned short port, unsigned char val)
+{
+    outb(val, port);
+}
+
+static __inline__ void
+xf_outw(unsigned short port, unsigned short val)
+{
+    outw(val, port);
+}
+
+static __inline__ void
+xf_outl(unsigned short port, unsigned int val)
+{
+    outl(val, port);
+}
+
+#define outb xf_outb
+#define outw xf_outw
+#define outl xf_outl
+
 #   else /* ix86 */
 
 #    define ldq_u(p)	(*((unsigned long  *)(p)))
