--- unix/vncviewer/CConn.cxx~	2006-09-03 22:28:17.819338250 +0200
+++ unix/vncviewer/CConn.cxx	2006-09-08 06:54:34.233326750 +0200
@@ -218,8 +218,8 @@
     if (user) {
       /* Get username */
       fprintf(stderr, "Username: ");
-      *user = new char[40];
-      fgets(*user, 40, stdin);
+      *user = new char[128];
+      fgets(*user, 128, stdin);
     }
 
     if (tcgetattr (fileno (stdin), &oldio) != 0) {
@@ -234,8 +234,8 @@
 	popup = 1;
       
       /* Read the password. */
-      *password = new char[8];
-      fgets (*password, 8, stdin);
+      *password = new char[64];
+      fgets (*password, 64, stdin);
 
       /* Restore terminal. */
       (void) tcsetattr (fileno (stdin), TCSAFLUSH, &oldio);
