--- unix/vncviewer/CConn.cxx.orig	2006-09-24 21:52:37.731592500 +0200
+++ unix/vncviewer/CConn.cxx	2006-09-24 21:52:42.575895250 +0200
@@ -220,6 +220,8 @@
       fprintf(stderr, "Username: ");
       *user = new char[128];
       fgets(*user, 128, stdin);
+      /* Remove \n at the end */
+      (*user)[strlen(*user)-1] = '\0';
     }
 
     if (tcgetattr (fileno (stdin), &oldio) != 0) {
@@ -236,6 +238,8 @@
       /* Read the password. */
       *password = new char[64];
       fgets (*password, 64, stdin);
+      /* Remove \n at the end */
+      (*password)[strlen(*password)-1] = '\0';
 
       /* Restore terminal. */
       (void) tcsetattr (fileno (stdin), TCSAFLUSH, &oldio);
