--- common/network/TcpSocket.cxx	2012-02-05 17:04:07 +0000
+++ common/network/TcpSocket.cxx	2012-02-05 17:06:16 +0000
@@ -260,6 +260,7 @@
   VNC_SOCKLEN_T addrlen = sizeof(struct sockaddr_in);
 
   getpeername(getFd(), (struct sockaddr *)&peeraddr, &addrlen);
+  addrlen = sizeof(struct sockaddr_in);
   getsockname(getFd(), (struct sockaddr *)&myaddr, &addrlen);
 
   return (peeraddr.sin_addr.s_addr == myaddr.sin_addr.s_addr);
