diff -ur vnc4-4.1.1+xorg4.3.0/common/network/TcpSocket.cxx vnc4-4.1.1+xorg4.3.0.new/common/network/TcpSocket.cxx
--- vnc4-4.1.1+xorg4.3.0/common/network/TcpSocket.cxx	2011-09-02 13:30:44.000000000 -0700
+++ vnc4-4.1.1+xorg4.3.0.new/common/network/TcpSocket.cxx	2011-09-01 18:17:07.074243662 -0700
@@ -282,18 +282,12 @@
   return true;
 }
 
-bool TcpSocket::isSocket(int sock)
+bool TcpSocket::isListening(int sock)
 {
-  struct sockaddr_in info;
-  VNC_SOCKLEN_T info_size = sizeof(info);
-  return getsockname(sock, (struct sockaddr *)&info, &info_size) >= 0;
-}
-
-bool TcpSocket::isConnected(int sock)
-{
-  struct sockaddr_in info;
-  VNC_SOCKLEN_T info_size = sizeof(info);
-  return getpeername(sock, (struct sockaddr *)&info, &info_size) >= 0;
+  int listening = 0;
+  VNC_SOCKLEN_T listening_size = sizeof(listening);
+  return getsockopt(sock, SOL_SOCKET, SO_ACCEPTCONN, &listening,
+                    &listening_size) >= 0 && listening;
 }
 
 int TcpSocket::getSockPort(int sock)
diff -ur vnc4-4.1.1+xorg4.3.0/common/network/TcpSocket.h vnc4-4.1.1+xorg4.3.0.new/common/network/TcpSocket.h
--- vnc4-4.1.1+xorg4.3.0/common/network/TcpSocket.h	2011-09-02 13:30:44.000000000 -0700
+++ vnc4-4.1.1+xorg4.3.0.new/common/network/TcpSocket.h	2011-09-01 18:17:25.524243654 -0700
@@ -59,8 +59,7 @@
     virtual void shutdown();
 
     static bool enableNagles(int sock, bool enable);
-    static bool isSocket(int sock);
-    static bool isConnected(int sock);
+    static bool isListening(int sock);
     static int getSockPort(int sock);
   private:
     bool closeFd;
diff -ur vnc4-4.1.1+xorg4.3.0/unix/xc/programs/Xserver/vnc/vncExtInit.cc vnc4-4.1.1+xorg4.3.0.new/unix/xc/programs/Xserver/vnc/vncExtInit.cc
--- vnc4-4.1.1+xorg4.3.0/unix/xc/programs/Xserver/vnc/vncExtInit.cc	2011-09-02 13:30:44.000000000 -0700
+++ vnc4-4.1.1+xorg4.3.0.new/unix/xc/programs/Xserver/vnc/vncExtInit.cc	2011-09-01 18:15:50.874243690 -0700
@@ -160,8 +160,7 @@
         network::TcpListener* listener = 0;
         network::TcpListener* httpListener = 0;
         if (scr == 0 && vncInetdSock != -1) {
-          if (network::TcpSocket::isSocket(vncInetdSock) &&
-              !network::TcpSocket::isConnected(vncInetdSock))
+          if (network::TcpSocket::isListening(vncInetdSock))
           {
             listener = new network::TcpListener(0, 0, vncInetdSock, true);
             vlog.info("inetd wait");
