--- buildtemp/xc/programs/Xserver/hw/xfree86/os-support/linux/lnx_io.c~	2001-11-08 05:00:14.000000000 +0100
+++ buildtemp/xc/programs/Xserver/hw/xfree86/os-support/linux/lnx_io.c	2004-08-09 09:02:19.000000000 +0200
@@ -80,19 +80,20 @@
    struct kbd_repeat kbdrep_s;
 
    /* don't change, just test */
-   kbdrep_s.rate = -1;
+   /*kbdrep_s.rate = -1;*/
    kbdrep_s.delay = -1;
    if (ioctl( 0, KDKBDREP, &kbdrep_s )) {
        return 0;
    }
 
    /* do the change */
-   if (rate == 0)				/* switch repeat off */
-     kbdrep_s.rate = 0;
-   else
-     kbdrep_s.rate  = 10000 / rate;		/* convert cps to msec */
-   if (kbdrep_s.rate < 1)
-     kbdrep_s.rate = 1;
+   /*if (rate == 0)*/				/* switch repeat off */
+   /*  kbdrep_s.rate = 0;*/
+   /*else*/
+   /*  kbdrep_s.rate  = 10000 / rate;*/		/* convert cps to msec */
+   /*  if (kbdrep_s.rate < 1)*/
+   /*  kbdrep_s.rate = 1; */
+
    kbdrep_s.delay = delay;
    if (kbdrep_s.delay < 1)
      kbdrep_s.delay = 1;
